messages/compiler/cpp/serialize.cpp

```diff
@@ -83,6 +83,23 @@ void deserialize(const uint8_t*& start, const uint8_t* end, T& t) {
   }
 }
 
+/******************************************************************************
+ * Enums
+ *
+ * Enums are type wrappers around uint8_t
+ ******************************************************************************/
+template <typename T, typename std::enable_if<std::is_enum<T>::value>::type* = nullptr>
+void serialize(std::vector<uint8_t>& output, const T& t) {
+  serialize(output, static_cast<uint8_t>(t));
+}
+
+template <typename T, typename std::enable_if<std::is_enum<T>::value>::type* = nullptr>
+void deserialize(const uint8_t*& start, const uint8_t* end, T& t) {
+  uint8_t val;
+  deserialize(start, end, val);
+  t = static_cast<T>(val);
```

Comment:
Does it make sense to check if the deserialised value is valid for type T (like we do for oneof)? It can be done with some C++ trickery or adding an ugly-named last option to the enum and checking the value against the enum range.

Reply:
It's a really good point, especially considering we do check this in python. I'm not sure we can just add an extra parameter due to recursive calls, but maybe we can. C++ trickery always an option. I'll think about it. Thanks for the review!

Reply:
What I meant was adding the option to all enums, something like:

```
enum class Color {
  red = 0,
  green = 1,
  blue = 2,
  __UGLY_LONG_AND_UNUSED_NAME_AS_LAST_OPTION = 3
};
```

and then compare for the [0, T::__UGLY_LONG_AND_UNUSED_NAME_AS_LAST_OPTION] range.

Reply:
Ah, yeah i did think of that. I believe I can do it with a function template specialization that returns the max size and then using that as a check in the deserialize function. I'm attempting that now.

Reply:
I added support for namespace scoped [`enumSize` functions](https://github.com/vmware/concord-bft/pull/1322/files#diff-2558f8429abbfe0d6f9d30ef7191bfc071b7edaaf043a62dd28bc85dc052794dR191-R192) and used [Argument Dependent Lookup](https://en.cppreference.com/w/cpp/language/adl) to ensure the [checks get called](https://github.com/vmware/concord-bft/pull/1322/files#diff-fdbbe49391692d69f4b4eb144885d711630f1f50ecbd5c12b3b754f2c8011e0cR101-R103). The argument itself gets ignored because the size is computed by the code generation, but it's needed for the ADL so arbitrary user namespaces can be looked in similar to our generated serialize functions for Msgs and oneofs.